src/estimators/stratified_sketch.py

```diff
@@ -0,0 +1,355 @@
+    )
+
+  def create_sketches(self):
+    if(self.sketches != {}):
+      return
+    reversedict = {}
+    for k, v in self.underlying_set.ids().items():
+      reversedict.setdefault(min(v, self.max_freq), []).append(k)
+
+    for freq in range(1, self.max_freq):
+      self.sketches[freq] = self.cardinality_sketch_factory(self.seed)
+      if (freq in reversedict):
+        self.sketches[freq].add_ids(reversedict[freq])
+
+    # Initialize max_freq
+    max_key = str(self.max_freq) + '+'
+    self.sketches[max_key] = self.cardinality_sketch_factory(self.seed)
+    if (self.max_freq in reversedict):
+      self.sketches[max_key].add_ids(reversedict[self.max_freq])
+
+  def _destroy_sketches(self):
+    self.sketches = {}
+
+  def add(self, x):
+    if (self.sketches != {}):
+      self._destroy_sketches()
+      logging.warn(
+          """Tried to add ids after sketch creation, sketches are destroyed.""",
+          RuntimeWarning)
+    self.underlying_set.add(x)
+
+  @classmethod
+  def init_from_exact_multi_set(cls, max_freq, exact_multi_set,
+                                cardinality_sketch_factory, random_seed):
+    """Initialize a Stratified sketch from one ExactMultiSet.
+
+    Args:
+      exact_multi_set: ExactMultiSet object to use for initialization.
+      CardinalitySketch: Class type of cardinality sketches this stratified
+        sketch will hold.
+    """
+    assert (cardinality_sketch_factory is
+            not None), ('cardinality_sketch is None')
+    stratified_sketch = cls(
+        max_freq=max_freq,
+        underlying_set=exact_multi_set,
+        cardinality_sketch_factory=cardinality_sketch_factory,
+        random_seed=random_seed)
+    stratified_sketch.create_sketches()
+
+    return stratified_sketch
+
+  @classmethod
+  def init_from_set_generator(cls, max_freq, set_generator,
+                              cardinality_sketch_factory, random_seed):
+    """Initialize a Stratified sketch from a Set Generator.
+
+    Args:
+      set_generator: SetGenerator object to draw ids from for initialization.
+      CardinalitySketch: Class type of cardinality sketches this stratified
+        sketch will hold.
+    """
+    assert (cardinality_sketch_factory is
+            not None), ('cardinality_sketch is None')
+    exact_multi_set = ExactMultiSet()
+    for generated_set in set_generator:
+      exact_multi_set.add_ids(generated_set)
+    return cls.init_from_exact_multi_set(max_freq, exact_multi_set,
+                                         cardinality_sketch_factory,
+                                         random_seed)
+
+  def assert_compatible(self, other):
+    """"Check if the two StratifiedSketch are comparable.
+
+    Args:
+      other: the other StratifiedSketch for comparison.
+
+    Raises:
+      AssertionError: if the other sketches are not StratifiedSketch, or if
+      their random_seed are different, or if the frequency targets are
+      different.
+    """
+    assert isinstance(other,
```

Comment:
Should they have the same cardinality sketch factory? What about underlying set type?

Reply:
asserted the types